x-pack/plugins/security/server/authentication/authenticator.test.ts

```diff
@@ -1318,6 +1356,130 @@ describe('Authenticator', () => {
+
+          mockOptions.session.get.mockResolvedValue({ error: failureReason, value: null });
+
+          await expect(authenticator.authenticate(request)).resolves.toEqual(
+            failureReason instanceof SessionMissingError
+              ? AuthenticationResult.notHandled()
+              : AuthenticationResult.failed(failureReason, {
+                  authResponseHeaders: {
+                    [SESSION_ERROR_REASON_HEADER]: (failureReason as SessionError).code,
+                  },
+                })
+          );
+          expect(auditLogger.log).not.toHaveBeenCalled();
+        });
+
+        it('should get expected reponse headers for non-redirectable requests where the authentication succeeds', async () => {
+          const request = httpServerMock.createKibanaRequest({ headers: { 'kbn-xsrf': 'xsrf' } });
+          const failureReason = new FailureClass();
+
+          mockOptions.session.get.mockResolvedValue({ error: failureReason, value: null });
+
+          const user = mockAuthenticatedUser();
+          mockBasicAuthenticationProvider.authenticate.mockResolvedValue(
+            AuthenticationResult.succeeded(user)
+          );
+
+          const authenticationResult = await authenticator.authenticate(request);
+          const expectedResult =
+            failureReason instanceof SessionExpiredError
+              ? AuthenticationResult.succeeded(user, {
+                  authResponseHeaders: {
+                    [SESSION_ERROR_REASON_HEADER]: (failureReason as SessionError).code,
+                  },
+                })
+              : AuthenticationResult.succeeded(user);
+          expect(authenticationResult).toEqual(expectedResult);
+          expect(auditLogger.log).not.toHaveBeenCalled();
+        });
+
+        it('should get expected reponse headers for non-redirectable requests where the authentication fails', async () => {
+          const request = httpServerMock.createKibanaRequest({ headers: { 'kbn-xsrf': 'xsrf' } });
+          const failureReason = new FailureClass();
+
+          mockOptions.session.get.mockResolvedValue({ error: failureReason, value: null });
+
+          const authError = new Error('foo');
+          mockHTTPAuthenticationProvider.authenticate.mockResolvedValue(
+            AuthenticationResult.failed(authError)
+          );
+
+          const authenticationResult = await authenticator.authenticate(request);
+          const expectedResult =
+            failureReason instanceof SessionExpiredError
+              ? AuthenticationResult.failed(authError, {
+                  authResponseHeaders: {
+                    [SESSION_ERROR_REASON_HEADER]: (failureReason as SessionError).code,
+                  },
+                })
+              : AuthenticationResult.failed(authError);
+          expect(authenticationResult).toEqual(expectedResult);
+          expectAuditEvents({ action: 'user_login', outcome: 'failure' }); // TODO: Do we really expect these audit events here?
```

Comment:
@thomheymann can you verify that this is expected?

Reply:
As discussed on zoom just now, we need to define what the expected behaviour should be for these cases since we're now treating "authentication not handled" as "authentication failed" in certain cases. 

The table in this PR describes the current audit logging logic and might provide pointers:
https://github.com/elastic/kibana/pull/139492#issue-1351148435

Reply:
Couple thoughts from a user perspective:

- `SessionExpiredError` (Session has expired and will be invalidated) - I don't think this is an authentication failure. We should probably treat this as `user_logout` the same way session cleanup is doing. 
- `SessionMissingError` (Session cookie does not exist) - What are the user scenarios when this could happen? 
- `SessionUnexpectedError ` - What are the user scenarios when this could happen?

Reply:
As agreed on zoom I've currently disabled the new audit log checks with a TODO to address this in a follow up PR: https://github.com/elastic/kibana/pull/139811/commits/9dd9b43823191046635f448c9274723ae9500884

Reply:
> - `SessionExpiredError` (Session has expired and will be invalidated) - I don't think this is an authentication failure. We should probably treat this as `user_logout` the same way session cleanup is doing. 

If I recall, we went with `failed` mainly because it was considered an error that the client hadn't auto-logged out the user before the session expired. This was done after discussing it with @azasypkin 

> - `SessionMissingError` (Session cookie does not exist) - What are the user scenarios when this could happen? 

This is actually the default behaviour if the user doesn't have a cookie - e.g. signing in for the first time - so fairly common and not something that warrants an error message being displayed.

> - `SessionUnexpectedError ` - What are the user scenarios when this could happen? 

This can happen if the session value is not available in the index or if the server is unable to decrypt session content (see `session.get` for implementation details).